### ISSUE TITLE
Add rentable storage

### ISSUE THREAD

**Issue by lock9** [neo-project/neo#1187]
**Summary**
We need a strategy for users to release space. One option is provide them the ability to rent the storage for a period of time/blocks.
This is common use case where after a certain period of time, we either expect the “transaction”(action) to be confirmed or to be considered expired (can be deleted)

**Do you have any solution you want to propose?**
We could have a cheaper storage option that automatically expires after some time. You pay by the space used and the time you want to retain your data.

**Where in the software does this update applies to?**
- Ledger
- SDK

**Comment by doubiliu** [comment]
Maybe we can use a smart contract to implement.This smart contract will have 3 base functions:
putStorage\getStorage\renewing

putStorage(String key,String value){
   In this process,user's data value will be store as a <K,V,S,E> type. K=prefix+key.V=value. (prefix can use tx sender's address).There is a another <K,S,E> type to store validity period.
S=startheight,E=endheight=startheight+default(10);
}

getStorage(String key){
   In this process,user can query the data by key.But before return the data,it will check the 
validity period first.If current height >endHeight,the data will be delete and not return.
}

renewing(String key,int height,byte[] from, BigInteger amount){
  In this process,user can pay  the time you want to retain  data.
}

**Comment by doubiliu** [comment]
what do you mean?The data in tx or the tx
If you mean tx store a data in storage
Maybe we can use a smart contract to implement.This smart contract will have 3 base functions:
putStorage\getStorage\renewing

putStorage(String key,String value){
In this process,user's data value will be store as a <K,V,S,E> type. K=prefix+key.V=value. (prefix can use tx sender's address).There is a another <K,S,E> type to store validity period.
S=startheight,E=endheight=startheight+default(10);
}

getStorage(String key){
In this process,user can query the data by key.But before return the data,it will check the
validity period first.If current height >endHeight,the data will be delete and not return.
}

renewing(String key,int height,byte[] from, BigInteger amount){
In this process,user can pay gas for the time you want to retain data.
}

**Comment by lock9** [comment]
@doubiliu  renewing is a good strategy, I haven't thought about that.
When you say do this using a smart contract, you mean having some kind of "smart service" that has this kind of storage?
I think we need more changes because this will impact the execution prices, but in any case, the renewal idea is great

**Comment by doubiliu** [comment]
@lock9 I mean we use a native contract.And we need corresponding Opcode or Interoperable service,like TempPut/TempGet

**Comment by lock9** [comment]
Hi @doubiliu and @neo-project/core.

One change  is needed:
- We can't delete data when it is queried, otherwise abandoned data will likely stay forever. We need to clean the storage every block (or, every few blocks - but I think we should do it every block). 

Also:
- Initially, I'm storing data using the block height as the prefix. When the user extends the storage, I move the value to the new expiration block height. 
- Using block index can be dangerous if we change the consensus. We need to support both block expiration height and expiration date. The second must be more expensive to use because it costs more to process. So far I only started logic on expiration using block height.
- I believe we can try to use temporary storage for state agreements. Since we have to update the storage every block to release storage, we can update the peers about our current state. I'm not going to do anything about this now.
- I think we need to group expiration dates by 5 or 10 seconds, otherwise, the number of keys will be too large.  How much do you guys think it is reasonable? Should we add this to the network policy? It seems the case, but I'm not sure.
- Is there any "protection" we should add to avoid incorrect usage of this feature? 

Do you have any feedback about the proposed solution? Thanks.

**Comment by shargon** [comment]
What is the use case for temporary information in a contract?

**Comment by lock9** [comment]
@shargon  If you have temporary storage, you can do some part of your logic using the temporary storage, and only move to the permanent storage when the data is "complete".

The user register on the SmartContract, but fails to pass KYC. What do you do?
- Pay an additional transaction to remove this data;
- Leave it there forever;
- Use temporary storage during this period, so if the user takes no action (or you don't approve it), data is automatically released;

It seems that the best solution is the third one.
In other words, rentable storage can be beneficial for the network, because it allows developers to release the storage without having to pay an additional fee or sending a new transaction, that would occupy even more storage.

**Comment by shargon** [comment]
who will make the transaction for review all of these storages than need to be cleaned?

**Comment by lock9** [comment]
@shargon  We either need to use "checkpoints" or do it every block. It has to be at "protocol level",  since it cannot depend on external actions.

I prefer to do it every block.

**Comment by Jim8y** [comment]
Sort of being fixed in other forms. And also inactive for 5 years.

**Comment by vncoelho** [comment]
I think it is still a good feature and nothing changed since then.

The issue is still updated and suitable to our context.

**Comment by roman-khimov** [comment]
Yes, this can still be considered in some form. #2034 is a related one to me (more specific, but maybe practically sufficient).
